### ISSUE TITLE
Enable remaining HTTP clients in metrics integration tests

### ISSUE THREAD

**Issue by scottgerring** [open-telemetry/opentelemetry-rust#2441]
in #2432 and the associated PR, we added integration tests for metrics.

 However,  HTTP-client exporters have issues that cause test failures, so they are skipped (we are only testing gRPC currently):

https://github.com/open-telemetry/opentelemetry-rust/blob/551760b220985c7c63a2b391eed797d26aff2327/opentelemetry-otlp/tests/integration_test/tests/metrics.rs#L186-L194

Remove this exclusion and fix the tests or the impl!

**Comment by scottgerring** [comment]
I will look at this next

**Comment by scottgerring** [comment]
The mystery deepens. I rolled my own futures-timeout that does not depend on tokio (yes, I know):

https://github.com/open-telemetry/opentelemetry-rust/blob/0254bc1c74320fb87d092995822ec13f871cd304/opentelemetry-http/src/lib.rs#L225-L256

This gets us further - we no longer panic in the timeout, but we panic further into hyper, where again we are expecting a tokio runtime - screenshot shows the frame that kicks back into tokio, and eventually panics for lack of a tokio runtime:

![CleanShot 2024-12-18 at 14 53 05@2x](https://github.com/user-attachments/assets/a4e20be7-626a-4484-8a50-b5cd830b1e75)

@cijothomas , I feel like I am missing something! It's not clear to me how all this rather-tokio-dependant-looking-stuff can run within a non-tokio thread context.  I can see that the path for the logging integration ends up with its own tokio context for hyper to work with, rather than sticking it on a non-tokio thread:

https://github.com/open-telemetry/opentelemetry-rust/blob/0254bc1c74320fb87d092995822ec13f871cd304/opentelemetry-http/src/lib.rs#L139-L147


Is this expected to not be working? I know we spoke about this the other day e.g. #1437 , but I suppose I assumed that the code already in `main` was already good.

**Comment by lalitb** [comment]
Will it make sense to document/support _ONLY_ below scenarios with dedicated-thread periodic reader:
- OTLP/gRPC - Using `tonic` client
- OTLP/HTTP - Using `reqwest-blocking` client

It seems that `hyper` client and `reqwest` client will only work with runtime-based periodic reader.

**Comment by scottgerring** [comment]
> Will it make sense to document/support _ONLY_ below scenarios with dedicated-thread periodic reader:
> 
>     * OTLP/gRPC - Using `tonic` client
> 
>     * OTLP/HTTP - Using `reqwest-blocking` client
> 
> 
> It seems that `hyper` client and `reqwest` client will only work with runtime-based periodic reader.

That would certainly make this ticket easier :)

**Comment by cijothomas** [comment]
For 1.0, our default is the dedicated thread and default gRPC client is Tonic, and default HTTP Client is reqwest::blocking, both of which are proven to be working with integration tests.

Reqwest/Hyper may have been designed to only work within tokio::runtimes, and we can just ship 1.0 without its support. (if anyone truly wants it, then the experimental-tokio-rt feature can be enabled to get it).

Post 1.0, we can see how to address supporting tonic/reqwest. I have some ideas on how to generally support other runtimes like tokio (and hence networking libs that depend on it like hyper), without exposing public APIs. Will post it into a new issue for discussion, but for 0.28 and 1.0 release, lets stick with @lalitb's [suggestion here](https://github.com/open-telemetry/opentelemetry-rust/issues/2441#issuecomment-2551493236)

**Comment by cijothomas** [comment]
> The mystery deepens. I rolled my own futures-timeout that does not depend on tokio (yes, I know):

@ThomsonTan also explored this to enforce timeouts within the dedicated thread. The general conclusion was that
_its best to let the individual exporters handle/enforce timeouts._ , rather than OTel trying to do the job of an executor!

We may revisit this in the future.

**Comment by scottgerring** [comment]
@cijothomas @lalitb perfect. I'll close this issue as a no-op for the moment!

**Comment by cijothomas** [comment]
Could you check and see if this can be added to the OTLP docs somehow, so users know what to expect. This can be done after we merge the Batchprocessor changes as well.

**Comment by scottgerring** [comment]
Sure - that's a good idea !

**Comment by scottgerring** [comment]
What if we also used the feature to remove the threaded `PeriodicReader` when the user selects an async HTTP client? I haven't checked if it has flow on effects, but this would save others from going down this rabbit hole.

**Comment by cijothomas** [comment]
Users can do it themselves by opting in to the feature flags. We should not do it ourselves, as that'd need using non-stable features.

I don't really think such issues will be common once Otel picks a default. Previously we didn't have a default, forcing user to pick one. Once we have a default, that we validate will work with one http and gRPC library, then users have less reason to change anything. If they need to change something, the doc can give advice about what combinations are invalid.

Defaults would be:
PeriodicReader, BatchProcessor that spawn own thread + HTTP + ReqwestBlockingClient.
Users can switch to gRPC+Tonic.

Anything else - requires opt-in to the experimental ff for custom-runtimes. Does this sound reasonable for 1.0 timeline?

**Comment by cijothomas** [comment]
> Thinking ahead, do we need 3 different async HTTP clients in this library at all 

I think just one would have been sufficient (reqwest-blocking, which works well in default mode of own background thread). But in asynchronous runtimes, we need non-blocking one, so either reqwest or hyper is needed. 

I am equally unsure of the need to support many clients! (Also ability to bring-your-own-http-client capability.. maybe there are use cases, I am happy to explore and write down the needs)

**Comment by scottgerring** [comment]
> > Thinking ahead, do we need 3 different async HTTP clients in this library at all
> 
> I think just one would have been sufficient (reqwest-blocking, which works well in default mode of own background thread). But in asynchronous runtimes, we need non-blocking one, so either reqwest or hyper is needed.
> 
> I am equally unsure of the need to support many clients! (Also ability to bring-your-own-http-client capability.. maybe there are use cases, I am happy to explore and write down the needs)

I suspect the use cases should focus on the customizations the user might need to provide - e.g. "I need {customHeaderX} for my use case, because of weird proxy stuff". Or, I need a callback hook _to provide_ a custom header, because its a bit more complicated than that". Then the problem becomes - which library makes it easier for us to provide the flexibility we might need in the future? I'm struggling to imagine a situation where a user of the library should need to BYO HTTP client; I think that would smell a bit like a break down of abstraction on our part. But - perhaps again here there is something that's come up in the other otel SDKs that shines light here?

@cijothomas - I feel like this is something that could use an RFC, and would be happy to start writing one if that fits :)

**Comment by cijothomas** [comment]
> I'm struggling to imagine a situation where a user of the library should need to BYO HTTP client; I think that would smell a bit like a break down of abstraction on our part. But - perhaps again here there is something that's come up in the other otel SDKs that shines light here?

My own experience is mostly .NET, where HttpClient is part of standard library itself, so this was not even a discussion point! (gRPC/protobuf was challenging, but ultimately Otel .NET decided to handcode a protobuf serializer, and got rid of gRPC/protobuf dependencies completely)

@lalitb you mentioned Otel C++ provides some ability to bring-own-http-client?

> @cijothomas - I feel like this is something that could use an RFC, and would be happy to start writing one if that fits?

Totally! One scenario I can think of is the security patches - where companies want to provide own http/gRPC clients, since Rust std library has none. (To be clear - even today, we don't support bring-you-own-client capability. we just allow user to pick between 3 implementations maintained by this repo itself.)

For a 1.0 priority, it is sufficient that we support tonic+reqwest-blocking , so we just need to make sure users have enough flexibility to control these libraries like adding extra headers or authentication etc. Post 1.0 is a good time to revisit if we need more flexibility to bring-own-http-client.

(A similar thing for async runtimes and bring-your-own-runtime ability. Good post 1.0 candidate)

**Comment by lalitb** [comment]
> @lalitb you mentioned Otel C++ provides some ability to bring-own-http-client?

Thats correct. C++ lacks a universally accepted standard HTTP client—`libcurl` is commonly used in POSIX environments, while Windows and macOS have their own native clients. Then there could be iOS and Android bindings on top of Otel C++ SDK which have their own http libraries - we can't impose the use of a specific HTTP client. Instead, need to provide flexible and pluggable setup to accommodate different implementations.

**Comment by cijothomas** [comment]
Can be closed now. Docs are already indicating which clients are supported vs not.
